### PR TITLE
PL/Python adding support for multi-dimensional arrays

### PR DESCRIPTION
**Pull request by davecramer**
Add support for multi-dimensional arrays
code from upstream to use recursion instead of iteration

### CODE REVIEW

**Comment by ivannovick** [comment]
cool stuff

**Comment by hlinnaka** [comment]
This looks nothing like what was committed upstream..

**Comment by hlinnaka** [comment]
Can you squash out and force push just the latest version of this patch, please? It's confusing to have the first bogus commit in the PR.

File: src/pl/plpython/expected/plpython_function.out [hunk 1]
```diff
@@ -158,67 +158,67 @@ CREATE TRIGGER show_trigger_data_trig
 BEFORE INSERT OR UPDATE OR DELETE ON trigger_test
 FOR EACH ROW EXECUTE PROCEDURE trigger_data(23,'skidoo');
 insert into trigger_test values(1,'insert');
-NOTICE:  TD[args] => ['23', 'skidoo']  (seg0 gxymac.local:40000 pid=35258)
+NOTICE:  TD[args] => ['23', 'skidoo']  (seg0 192.168.1.138:25432 pid=36868)
```
Comment (hlinnaka) [inline src/pl/plpython/expected/plpython_function.out]: Let's avoid the unnecessary churn of these changes that are masked by gpdiff anyway.

File: src/pl/plpython/expected/plpython_function.out [hunk 2]
```diff
@@ -540,24 +540,7 @@ $$ LANGUAGE plpythonu;
 CREATE FUNCTION test_inout_params(first inout text) AS $$
 return first + '_inout';
 $$ LANGUAGE plpythonu;
-CREATE FUNCTION test_type_conversion_bool(x bool) returns bool AS $$ return x $$ language plpythonu;
-CREATE FUNCTION test_type_conversion_char(x char) returns char AS $$ return x $$ language plpythonu;
```
Comment (hlinnaka) [inline src/pl/plpython/expected/plpython_function.out]: Why are all these being removed?

**Comment by davecramer** [comment]
Because they are moved to plpython_types which is more like how upstream
does it

Dave Cramer

File: src/pl/plpython/expected/plpython_function.out (hunk 2 above)
Comment (hlinnaka) [inline src/pl/plpython/expected/plpython_function.out]: Ah, I see, they're moved to the plpython_types test.

I think it'd be good to separate these refactorings to a separate commit, and apply the multi-dimensional patches on top of that.

**Comment by hlinnaka** [comment]
Please also add details to the commit message, to explain what the commit(s) do, why, and which upstream commits they correspond to.

**Comment by davecramer** [comment]
I've created another PR https://github.com/greenplum-db/gpdb/pull/1292

**Comment by hlinnaka** [comment]
I assume we don't need this PR anymore, because #1292 is more current. Closing.
